### PR TITLE
Multisubmit rules

### PR DESCRIPTION
**Pull request by lumber1000**

### CODE REVIEW

File: src/main/proto/th2_grpc_check1/check1.proto [hunk 1]
```diff
@@ -23,13 +23,13 @@ import "th2_grpc_common/common.proto";
...
 option java_package = "com.exactpro.th2.check1.grpc";
 
-
 service Check1 {
     rpc createCheckpoint (CheckpointRequest) returns (CheckpointResponse) {}
     rpc submitCheckRule (CheckRuleRequest) returns (CheckRuleResponse) {}
     rpc submitCheckSequenceRule(CheckSequenceRuleRequest) returns (CheckSequenceRuleResponse) {}
     rpc submitNoMessageCheck(NoMessageCheckRequest) returns (NoMessageCheckResponse) {}
     rpc waitForResult(WaitForResultRequest) returns (WaitForResultResponse) {}
+    rpc multiSubmitRules(MultiSubmitRulesRequest) returns (MultiSubmitRulesResponse) {}
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/proto/th2_grpc_check1/check1.proto:32]: Please rename to submitMultipleRules(MultiRulesRequest) returns (MultiRulesResponse)

File: src/main/proto/th2_grpc_check1/check1.proto [hunk 2]
```diff
@@ -132,4 +132,27 @@ message WaitForResultRequest {
...
+}
+
+message MultiSubmitRulesRequest {
+    repeated Rule rules = 1;
+    bool omit_response = 2;
+}
+
+message MultiSubmitRulesResponse {
+    repeated RuleResponse responses = 1;
+}
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/proto/th2_grpc_check1/check1.proto:144]:
```suggestion
message MultiRulesResponse {
    ChainID chain_id = 1; 
    repeated RuleResponse responses = 2;
}
```

File: src/main/proto/th2_grpc_check1/check1.proto [hunk 3]
```diff
@@ -132,4 +132,27 @@ message WaitForResultRequest {
...
+        CheckRuleRequest checkRuleRequest = 1;
+        CheckSequenceRuleRequest sequenceRule = 2;
+        NoMessageCheckRequest noMessageRule = 3;
+    }
+}
+
+message RuleResponse {
+    RequestStatus status = 1;
+    ChainID chain_id = 2; // You can use it to unite the next rule to the chain with currently submitted one
+    int64 rule_id = 3; // Can be used for requesting response via 'waitForResult' method
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/proto/th2_grpc_check1/check1.proto:157]:
```suggestion
message RuleResponse {
    RequestStatus status = 1;
    int64 rule_id = 2; // Can be used for requesting response via 'waitForResult' method
```

**Review by Nikita-Smirnov-Exactpro — CHANGES_REQUESTED** [review]

**Comment by Nikita-Smirnov-Exactpro** [comment]
I have questions about chain id in submit multiple rules. 
1. Should check1 provide ability to submit rules to different chain id? It is related to chain id in each request
2. Should check1 create separate chain id for each rules? It is related to chain id in response

My opinion is check1 should handle the whole multiple rules request as request for one chain id. This behavior should cover test case needs, because as I know users doesn't use chain id or use fixed chain id for case.
@OptimumCode what do you think about these questions?

**Comment by lumber1000** [comment]
Currently check1 allows to submit to one chain or to multiple chains or to generate new chainId for every rule. Demo script uses both fixed chainId for some rules and generating new chainId in check1 for others.

**Comment by OptimumCode** [comment]
> My opinion is check1 should handle the whole multiple rules request as request for one chain id. This behavior should cover test case needs, because as I know users doesn't use chain id or use fixed chain id for case. @OptimumCode what do you think about these questions?

I don't think it will look good with the current API state. Each rule request contains ChainID. So, it is expected that this chain ID will be used when the rule is submitted. At least I would expect that. If we apply your suggestion @Nikita-Smirnov-Exactpro the above won't be true and will look like an unexpected behavior for the user in my opinion.
We could introduce a top-level chain ID in MultiRulesRequest that could be used for all rules if a user didn't specify a chain ID for a particular rule. The same might be applied to the MultiRulesResponse, however, we should not remove the chain ID from the individual response per rule as some rules might have their own chain IDs.

With the current API, I would rather stick to the option the @lumber1000 used in the beginning.

**Comment by Nikita-Smirnov-Exactpro** [comment]
> We could introduce a top-level chain ID in MultiRulesRequest that could be used for all rules if a user didn't specify a chain ID for a particular rule. The same might be applied to the MultiRulesResponse, however, we should not remove the chain ID from the individual response per rule as some rules might have their own chain IDs.

I like this option, in this case user can submit multiple rules under single chain.
Also, I would like to use common chain id as priority way, for example
```
top-level-chain-id: empty
  rule1-chain-d: empty
  rule2-chain-d: chain-a
  rule3-chain-d: empty
```
check1 generates new `chain-0` chain id and submits 1,3 rule under it. Responses for each rule has actual chain id

File: src/main/proto/th2_grpc_check1/check1.proto [hunk 4]
```diff
@@ -132,4 +134,27 @@ message WaitForResultRequest {
 message WaitForResultResponse {
     RequestStatus status = 1;
     EventStatus rule_result = 2;
+}
+
+message MultiRulesRequest {
+    ChainID default_chain = 1;
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/proto/th2_grpc_check1/check1.proto:140]: Could we declare the common parent_event_id with the same logic:
each rule without parent event id takes the parent_event_id from rule request.

**Review by Nikita-Smirnov-Exactpro — APPROVED** [review]

**Review by Nikita-Smirnov-Exactpro — APPROVED** [review]

**Pull request merged**